hail/src/main/scala/is/hail/rvd/RVDPartitioner.scala

```diff
@@ -184,9 +194,10 @@ class RVDPartitioner(
 
   def copy(
     kType: TStruct = kType,
-    rangeBounds: IndexedSeq[Interval] = rangeBounds
+    rangeBounds: IndexedSeq[Interval] = rangeBounds,
+    allowedOverlap: Int = allowedOverlap
   ): RVDPartitioner =
-    new RVDPartitioner(kType, rangeBounds)
+    new RVDPartitioner(kType, rangeBounds, allowedOverlap)
```

Comment:
This `allowedOverlap` and the one in `coarsen` should be removed, so that `allowedOverlap` doesn't get added as a field.

Reply:
> so that `allowedOverlap` doesn't get added as a field

I removed `allowedOverlap` as a field (that is, it is no longer a `val`).  `copy` is a convenience constructor.  Shouldn't have the same arguments as the constructor and preserve the arguments that aren't overridden?

Isn't `coarsen` preserving the overlap (up to truncation)?  Why shouldn't it correctly propagate that information forward?

Maybe I don't know what you mean by "allowedOverlap" is a field here.

Reply:
I just mean that if `allowedOverlap` is only used in the (implicitly defined) constructor, then Scala doesn't need to store it in a field in the generated Java class.

`allowedOverlap` doesn't have any information content that needs to be propagated. Either the `rangeBounds` satisfy the property or they don't. Including `allowedOverlap` in `copy` and `coarsen` is just requesting an extra check during construction of the new partitioner that is impossible to fail. My intention for leaving `allowedOverlap` as a constructor argument was purely as a more convenient alternative to
```
assert(RVDPartitioner.isValid(kType, rangeBounds, allowedOverlap))
val partitioner = new RVDPartitioner(kType, rangeBounds)
```

Reply:
Ah, you mean you don't want to reference it outside the constructor in which case it will actually get stored?

I think I'm starting to see your logic.  `allowedOverlap` is so clients outside RVDPartitioner can specify the expected allowed overlap and get it checked?  I guess I have two questions: (1) Does that ever happen?  Who knows their allowedOverlap statically in an interesting way?  And (2) Why are you including internal uses of the RVDPartitioner constructor?  Shouldn't they get the benefit of that checking, too?  (Like coarsen?)

Reply:
> Including `allowedOverlap` in `copy` and `coarsen` is just requesting an extra check during construction of the new partitioner that is impossible to fail.

Unless there is a bug in coarsen or the caller of copy.  Isn't that exactly the kind of thing `allowedOverlap` is there to check for?

Reply:
I don't think it would take much to convince me to remove it completely. I left it there as a bridge to the old `partitionKeys` world.

1) As an example, `LoadVCF` calls `new RVDPartitioner(Array("locus"), fullType.rowKeyStruct, rangeBounds.asInstanceOf[IndexedSeq[Interval]])`, which dispatches to the primary constructor using the `allowedOverlap` argument to enforce that the partitioner being built is consistent with the old "partitionKey = locus" semantics.
2) Internal uses are easier to prove corectness. For example, if `rangeBounds` satisfies `allowedOverlap = n`, then of course it still will in the new partitioner made by `copy`. Whatever `allowedOverlap` should be passed to the constructor in `coarsen` should likewise be provably guaranteed to pass. Maybe storing `allowedOverlap` in the java object in order to get that extra runtime sanity check is worth it. I was less concerned about whether `allowedOverlap` gets stored than I was making clear that it is not semantically an attribute of the partitioner.

Reply:
I've never understood "impossible to fail" as a complaint against assertions.  All asserts should be impossible to fail by definition.  It's only the bugs that makes them not so, and that's what you're looking for.

Anyway, I don't really care to fight about this.  With `satisfiesAllowedOverlap`, my case is handled.  I'll back out the changes if you prefer.

Reply:
> Unless there is a bug in coarsen or the caller of copy. Isn't that exactly the kind of thing `allowedOverlap` is there to check for?

Yeah, you've convinced me. Ignore this request.

Reply:
Sorry, I didn't want a fight either. Just trying to explain my original reasoning. You're completely right about assertions.

Reply:
OK, great.

 > I was less concerned about whether `allowedOverlap` gets stored than I was making clear that it is not semantically an attribute of the partitioner.

So it feels like a semantic attribute to me, it is just a (very cheap!)  bound on the property you can compute exactly from rangeBounds in O(N).  But I see now where you were coming from.  Thanks for discussing!